float32[] p_1
float32[] p_2
float32[] p_L